field/src/extension/cubic.rs

```diff
@@ -0,0 +1,302 @@
+use core::fmt::{self, Debug, Display, Formatter};
+use core::iter::{Product, Sum};
+use core::ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Neg, Sub, SubAssign};
+
+use crate::extension::{Extendable, Frobenius, OEF};
+use crate::field::Field;
+use crate::{AbstractExtensionField, AbstractField, ExtensionField};
+
+#[derive(Copy, Clone, Eq, PartialEq, Hash)]
+pub struct CubicExtension<F: Extendable<3>>(pub [F; 3]);
+
+impl<F: Extendable<3>> Default for CubicExtension<F> {
+    fn default() -> Self {
+        Self::ZERO
+    }
+}
+
+impl<F: Extendable<3>> From<F> for CubicExtension<F> {
+    fn from(x: F) -> Self {
+        Self([x, F::ZERO, F::ZERO])
+    }
+}
+
+impl<F: Extendable<3>> AbstractField for CubicExtension<F> {
+    const ZERO: Self = Self([F::ZERO; 3]);
+    const ONE: Self = Self([F::ONE, F::ZERO, F::ZERO]);
+    const TWO: Self = Self([F::TWO, F::ZERO, F::ZERO]);
+    const NEG_ONE: Self = Self([F::NEG_ONE, F::ZERO, F::ZERO]);
+
+    #[inline(always)]
+    fn square(&self) -> Self {
+        let Self([a0, a1, a2]) = *self;
+        let w = <Self as OEF<F>>::W;
+
+        let c0 = a0.square() + w * a1 * a2;
+        let c1 = w * a2.square() + a0 * a1;
+        let c2 = a1.square() + a0 * a2;
+
+        Self([c0, c1, c2])
+    }
+
+    fn from_bool(b: bool) -> Self {
+        F::from_bool(b).into()
+    }
+
+    fn from_canonical_u8(n: u8) -> Self {
+        F::from_canonical_u8(n).into()
+    }
+
+    fn from_canonical_u16(n: u16) -> Self {
+        F::from_canonical_u16(n).into()
+    }
+
+    fn from_canonical_u32(n: u32) -> Self {
+        F::from_canonical_u32(n).into()
+    }
+
+    /// Convert from `u64`. Undefined behavior if the input is outside the canonical range.
+    fn from_canonical_u64(n: u64) -> Self {
+        Self::from_canonical_u32(
+            n.try_into()
+                .expect("Too large to be a canonical Mersenne31 encoding"),
+        )
+    }
+
+    /// Convert from `usize`. Undefined behavior if the input is outside the canonical range.
+    fn from_canonical_usize(n: usize) -> Self {
+        Self::from_canonical_u32(
+            n.try_into()
+                .expect("Too large to be a canonical Mersenne31 encoding"),
+        )
+    }
+
+    fn from_wrapped_u32(n: u32) -> Self {
+        F::from_wrapped_u32(n).into()
+    }
+
+    fn from_wrapped_u64(n: u64) -> Self {
+        F::from_wrapped_u64(n).into()
+    }
+
+    fn multiplicative_group_generator() -> Self {
+        F::multiplicative_group_generator().into()
```

Comment:
I think we want a generator of the whole EF's multiplicative group, with order `|F|^2 - 1`. Probably would need to be specified in `Extendable` as in Plonky2.

Reply:
fixed